libyul/interpreter/Scope.h

```diff
@@ -0,0 +1,68 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libyul/interpreter/types.h>
+
+#include <libyul/ASTForward.h>
+#include <libyul/YulName.h>
+
+#include <libsolutil/Numeric.h>
+
+#include <map>
+#include <vector>
+
+namespace solidity::yul::interpreter
+{
+
+/**
+ * Scope structure built and maintained during execution.
+ */
+class Scope
+{
+public:
+	Scope(Scope* const _parent = nullptr):
+		m_definedFunctions(),
+		m_declaredVariables(),
+		m_subScopes(),
+		m_parent(_parent)
+	{}
+
+	Scope* parent() const
+	{
+		return m_parent;
+	}
+
+	Scope* getSubscope(Block const& _block);
+
+	void addDeclaredVariable(YulName const& _name);
+
+	/// Note: m_declaredVariables will be cleared after this function
+	void cleanupVariables(VariableValuesMap& _variables);
+
+	/// Throw an error if function with the given name is not found.
+	FunctionDefinition const& getFunction(YulName const& _functionName) const;
+
+private:
+	std::map<YulName, FunctionDefinition const&> m_definedFunctions;
+	std::vector<YulName> m_declaredVariables;
+	std::map<Block const*, std::unique_ptr<Scope>> m_subScopes;
+	Scope* const m_parent;
```

Comment:
Our convention now is to always initialize variables of POD types in declaration (this might not have been the case when the original interpreter was written).

We do that even if they get a value assigned later in constructor - in that case I'd just use the default initialization with `{}` to emphasize that the value used here does not matter, only that it does get some deterministic value.

```suggestion
	Scope* const m_parent{};
```

You could use the same style with `PureInterpreter::m_recursionDepth`.

Reply:
I have put default initialization to fields as advised.